pkg/flock/flock_unix.go

```diff
@@ -0,0 +1,176 @@
+// build unix
+
+package flock
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/pkg/errors"
+	"golang.org/x/sync/semaphore"
+	"golang.org/x/sys/unix"
+)
+
+// New returns a Flock for path.
+func New(path string) (flock, error) {
+	// Make sure we can operate on the path.
+	fd, err := unix.Open(path, os.O_RDWR|os.O_CREATE, unix.S_IRUSR|unix.S_IWUSR)
+	if err != nil {
+		return flock{}, err
+	}
+	if err := unix.Close(fd); err != nil {
+		return flock{}, err
+	}
+	return flock{
+		flockType: flockUnlocked,
+		fd:        -1,
+		path:      path,
+		mutex:     new(sync.Mutex),
+		sem:       semaphore.NewWeighted(1),
+	}, nil
+}
+
+// open opens the flock's path and sets the file desriptor. State must be
+// synchronized by the caller.
+func (f *flock) open() error {
+	if f.fd != -1 || f.flockType != flockUnlocked {
+		return fmt.Errorf("%v: trying to open improper Flock %q %d %d", errInternal, f.path, f.fd, f.flockType)
```

Comment:
please use `errors.Wrapf`

Reply:
Initially I had that but the problem is that Wrapf mixes the order. I want to have a "internal error ..." but using Wrapf would move that to the end.

Reply:
Perhaps `errors.WithStack()`?